File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/converter/SyntheticSharesGenerator.java
package com.example.tradelog.api.converter;

import com.example.common.converter.TimeConverter;
import com.example.tradelog.api.spec.model.ActionType;
import com.example.tradelog.api.spec.model.ShareJournalModel;
import com.example.tradelog.api.spec.model.TransactionModel;
import com.example.tradelog.api.spec.model.TransactionSettingsModel;
import com.example.tradelog.api.spec.model.TransactionType;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

public class SyntheticSharesGenerator {

    public static Function<List<ShareJournalModel>, List<ShareJournalModel>> createSynthetic = shareJournalModels -> {
        Map<String, ShareAggregator> stocks = new HashMap<>();

        shareJournalModels.stream()
                .filter(f -> f.getTransactionDetails().getType().equals(TransactionType.SHARE))
                .filter(f -> !f.getTransactionDetails().getSettings().getLegClosed())
                .forEach(s -> {
                    int quantity;
                    if (ActionType.BUY == s.getAction()) {
                        quantity = s.getQuantity();
                    } else {
                        quantity = s.getQuantity() * -1;
                    }

                    ShareAggregator aggregator;
                    if (stocks.containsKey(s.getTransactionDetails().getSymbol())) {
                        aggregator = stocks.get(s.getTransactionDetails().getSymbol());
                    } else {
                        aggregator = new ShareAggregator(s.getTransactionDetails().getSymbol());
                        aggregator.setActualPrice(s.getActualPrice());
                        double preferredPrice = s.getTransactionDetails().getSettings().getPreferredPrice();
                        aggregator.setPreferredPrice(preferredPrice);
                    }

                    aggregator.addQuantityAndPrice(quantity, s.getPrice());
                    stocks.put(s.getTransactionDetails().getSymbol(), aggregator);
                });

        List<ShareJournalModel> synthetics = new ArrayList<>();
        stocks.forEach( (s, aggregator) -> {
            if (aggregator.getQuantity() != 0) {

                double averageBoughtPrice = aggregator.getAverageBoughtPrice();

                ActionType syntheticActionType = ActionType.SELL;
                if (aggregator.getQuantity() < 0) {
                    syntheticActionType = ActionType.BUY;
                }

                TransactionSettingsModel optionsModel = new TransactionSettingsModel("", aggregator.getPreferredPrice(), true, false);

                TransactionModel transactionModel = new TransactionModel(
                        "",
                        "",
                        TimeConverter.nextYear(),
                        s,
                        TransactionType.SYNTHETIC_SHARE,
                        0.00,
                        optionsModel);


                synthetics.add(new ShareJournalModel(transactionModel, averageBoughtPrice, aggregator.getActualPrice(),
                        Math.abs(aggregator.getQuantity()), syntheticActionType));
            }
        });

        return synthetics;

    };

}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/core/util/SyntheticSharesGenerator.kt
package com.example.tradelog.api.core.util

import com.example.common.converter.TimeConverter
import com.example.tradelog.api.core.model.*
import kotlin.math.abs

fun createSynthetic(models: List<ShareJournalModel>): List<ShareJournalModel> {
    val stockMap = HashMap<String, ShareAggregator>()

    models.stream()
            .filter {it.transactionDetails.type == TransactionType.SHARE}
            .filter {!it.transactionDetails.settings.legClosed}
            .forEach {
                val quantity = if (ActionType.BUY === it.action) {
                    it.quantity
                } else {
                    it.quantity * -1
                }

                val aggregator = if (stockMap.containsKey(it.transactionDetails.symbol)) {
                    stockMap[it.transactionDetails.symbol]!!
                } else {
                    ShareAggregator(symbol = it.transactionDetails.symbol,
                            actualPrice = it.actualPrice,
                            preferredPrice = it.transactionDetails.settings.preferredPrice)
                }

                aggregator.addQuantityAndPrice(quantity, it.price)
                stockMap[it.transactionDetails.symbol] = aggregator
            }

    val synthetics = ArrayList<ShareJournalModel>()
    stockMap.forEach {(symbol, aggregator) ->
        if (aggregator.quantity != 0) {
            val averageBoughPrice = aggregator.averageBoughtPrice
            var syntheticActionType = ActionType.SELL
            if (aggregator.quantity < 0) {
                syntheticActionType = ActionType.BUY
            }

            val settingsModel = TransactionSettingsModel(transactionId = "",
                    preferredPrice = aggregator.preferredPrice,
                    groupSelected = true, legClosed = false)

            val transactionModel = TransactionModel(id = "",
                    accountId = "",
                    date = TimeConverter.nextYear(),
                    symbol = symbol,
                    type = TransactionType.SYNTHETIC_SHARE,
                    brokerFees = 0.0,
                    settings = settingsModel)

            synthetics.add(ShareJournalModel(transactionDetails = transactionModel,
                    price = averageBoughPrice,
                    actualPrice = aggregator.actualPrice,
                    quantity = abs(aggregator.quantity),
                    action = syntheticActionType
            ))

        }
    }

    return synthetics
}